multiwallet.go

```diff
@@ -0,0 +1,403 @@
+			syncedWallets++
+		}
+	}
+
+	return syncedWallets
+}
+
+func (mw *MultiWallet) CreateNewWallet(privatePassphrase string, privatePassphraseType int32) (*LibWallet, error) {
+
+	if mw.activeSyncData != nil {
+		return nil, errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	seed, err := GenerateSeed()
+	if err != nil {
+		return nil, err
+	}
+
+	properties := &Wallet{
+		Seed:                  seed,
+		PrivatePassphraseType: privatePassphraseType,
+		DiscoveredAccounts:    true,
+	}
+
+	return mw.createWallet(properties, seed, privatePassphrase)
+}
+
+func (mw *MultiWallet) CreateWatchOnlyWallet(walletName string, extendedPublicKey string) (*LibWallet, error) {
+
+	exists, err := mw.WalletNameExists(walletName)
+	if err != nil {
+		return nil, err
+	} else if exists {
+		return nil, errors.New(ErrNotExist)
+	}
+
+	w := &Wallet{
+		Name:               walletName,
+		DiscoveredAccounts: true,
+	}
+
+	libWallet, err := mw.saveWalletToDatabase(w)
+	if err != nil {
+		return nil, err
+	}
+
+	err = libWallet.CreateWatchingOnlyWallet(wallet.InsecurePubPassphrase, extendedPublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	go mw.listenForTransactions(libWallet)
+
+	return libWallet, nil
+}
+
+func (mw *MultiWallet) RestoreWallet(seedMnemonic, privatePassphrase string, privatePassphraseType int32) (*LibWallet, error) {
+	if mw.activeSyncData != nil {
+		return nil, errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	properties := &Wallet{
+		PrivatePassphraseType: privatePassphraseType,
+		DiscoveredAccounts:    false,
+	}
+
+	return mw.createWallet(properties, seedMnemonic, privatePassphrase)
+}
+
+func (mw *MultiWallet) createWallet(w *Wallet, seedMnemonic, privatePassphrase string) (*LibWallet, error) {
+
+	libWallet, err := mw.saveWalletToDatabase(w)
+	if err != nil {
+		return nil, err
+	}
+
+	err = libWallet.CreateWallet(privatePassphrase, seedMnemonic)
+	if err != nil {
+		return nil, err
+	}
+
+	go mw.listenForTransactions(libWallet)
+
+	return libWallet, nil
+}
+
+func (mw *MultiWallet) saveWalletToDatabase(w *Wallet) (*LibWallet, error) {
+	// saving struct to update ID property with an autogenerated value
+	err := mw.db.Save(w)
+	if err != nil {
+		return nil, err
+	}
+
+	// delete from database if not created successfully
+	defer func() {
+		if err != nil {
+			mw.db.DeleteStruct(w)
+		}
+	}()
+
+	walletDataDir := filepath.Join(mw.rootDir, strconv.Itoa(w.ID))
+	os.MkdirAll(walletDataDir, os.ModePerm) // create wallet dir
+
+	w.DataDir = walletDataDir
+	err = mw.db.Save(w) // update database with complete wallet information
+	if err != nil {
+		return nil, err
+	}
+
+	w.walletDbDriver = mw.dbDriver
+	w.netType = mw.activeNet.Name
+	libWallet, err := NewLibWallet(w)
+	if err != nil {
+		return nil, err
+	}
+
+	mw.libWallets[w.ID] = libWallet
+	return libWallet, nil
+}
+
+func (mw *MultiWallet) WalletNameExists(walletName string) (bool, error) {
+
+	if strings.HasPrefix(walletName, "wallet-") {
+		return false, errors.E(ErrReservedWalletName)
```

Comment:
As a continuation of the conversation here https://github.com/raedahgroup/dcrlibwallet/pull/57#discussion_r344599402, I need more clarification from you.

Remember this method does not save the name to the db, it only checks if it exists. I believe this check is only necessary when attempting to save a name to db, then you can say "oh no, we can't save that name because it is reserved".

Reply:
The method checks if a passed wallet name is valid for use.